DEV: FT.AGGREGATE expression precedence (#1589)
redis/docs: content/develop/interact/search-and-query/advanced-concepts/aggregation-syntax.md

---
categories:
- docs
- develop
- stack
- oss
- rs
- rc
- oss
- kubernetes
- clients
description: Order of operations for the FT.AGGREGATE command
linkTitle: Aggregation syntax
title: FT.AGGREGATE order of operations
weight: 2
---

## Overview

[`FT.AGGREGATE`]({{< relref "/commands/ft.aggregate" >}}) is a powerful Redis Query Engine (RQE) command for performing advanced data aggregation, filtering, sorting, and transformations on indexed hash or JSON documents. This reference page provides a structured breakdown of syntax, ordering rules, and best practices.

> The [main aggregations page]({{< relref "/develop/interact/search-and-query/advanced-concepts/aggregations" >}}) has a simple diagram showing how `FT.AGGREGATE` pipelines are constructed, but it doesn't tell the whole story. For example, it's possible to create more complex aggregation pipelines by applying multiple `REDUCE` functions under a single `GROUPBY` clause, or you can chain groupings and mix in additional mapping steps:

Reviewer: ```suggestion
The [main aggregations page]({{< relref "/develop/interact/search-and-query/advanced-concepts/aggregations" >}}) has a simple diagram showing how `FT.AGGREGATE` pipelines are constructed, but it doesn't tell the whole story. For example, you can create more complex aggregation pipelines by applying multiple `REDUCE` functions under a single `GROUPBY` clause, or you can chain groupings and mix in additional mapping steps:
```

`GROUPBY` ... `REDUCE` ... `APPLY` ... `GROUPBY` ... `REDUCE`

{{< note >}}
The examples on this page are based on a hypothetical "products" data set, which you can [download here](./data/products.txt).
{{< /note >}}

## Syntax and expression ordering

The `FT.AGGREGATE` command processes multiple expressions in a pipeline. Below is the recommended order:

> 1. `index` – the name of your index, which must be the first argument.

Reviewer 2: All numbers are `1`

Author: Hehe... I get that a lot. When Hugo converts Markdown to HTML, it re-numbers the list in sequential order. Using all ones is a shortcut for writers when we have to add more items to an existing list.

1. `query` – your query, which must be the second argument.
1. `FILTER` – filters raw documents before transformations or aggregation.

> 1. `LOAD` – loads additional document attributes.

Reviewer 2: ```suggestion
1. `LOAD` – loads document fields.
```

1. `APPLY` – applies transformations on fields.
1. `GROUPBY` – groups results by specific fields.
1. `REDUCE` – performs aggregations. For example, `SUM`, `COUNT`, and `AVG`.
1. `SORTBY` – orders the results based on specified fields.
1. `LIMIT` – restricts the number of results returned.

> 1. `DIALECT 2` - provides for more comprehensive syntax, for example using parameters in `FILTER` expressions.

Reviewer 2: The content of the `DIALECT 2` bullet

> 1. `DIALECT 2` - provides for more comprehensive syntax, for example using parameters in `FILTER` expressions.

Reviewer 2: I would emphasize the changes to the **query** syntax more, since these are the more impactful changes.

Other keywords will be discussed toward the end of this page.

## When to use `@`

> Fields must be preceded by `@` in the following circumstances:

Reviewer: There's quite a lot of PV in the following bits (some valid because it's the syntax rather than the user that specifies what happens). With this one, I'm not sure if it works best as AV or PV:
```suggestion
You must add `@` at the start of a field name in the following circumstances:
```

> - When referencing fields loaded from documents. In the following example, `price` is a document field and must be prefixed wit `@`.

Reviewer: ```suggestion
- When referencing fields loaded from documents. In the following example, `price` is a document field and must be prefixed with `@`.
```

```sh
FT.AGGREGATE products "*"
  LOAD 1 @rating
  FILTER "@rating >= 4.5"
  LIMIT 0 10
```

- When referencing fields inside `GROUPBY` or `REDUCE` clauses.

```sh
FT.AGGREGATE products "*"
  GROUPBY 1 @category
  REDUCE SUM 1 @price AS total_price
  LIMIT 0 10
```

- When referencing fields created by `REDUCE` in an `APPLY` or `FILTER` clauses.

```sh
FT.AGGREGATE products "*"
  GROUPBY 1 @category
  REDUCE SUM 1 @price AS total_price
  APPLY "@total_price * 1.2" AS boosted_price
  FILTER "@total_price > 1000"
  LIMIT 0 10
```

- When referencing fields created by `APPLY` in another `APPLY` or `FILTER` clause.

```sh
FT.AGGREGATE products "*"
  LOAD 2 @price @discount
  APPLY "@price - @discount" AS net_price
  APPLY "@net_price * 1.1" AS marked_up
  FILTER "@net_price > 200"
  LIMIT 0 10
```

- When referencing fields created by `APPLY` in a `SORTBY` clause.

```sh
FT.AGGREGATE products "*"
  LOAD 2 @price @discount
  APPLY "@price - @discount" AS net_price
  SORTBY 2 @net_price DESC
  LIMIT 0 10
```

## GROUPBY with multiple REDUCE operations

`GROUPBY` can be followed by multiple `REDUCE` operations for different aggregations.

```sh

> FT.AGGREGATE products "*"

Reviewer 2: I think it would be nice to show the results of the queries you mention here, so that it is clearer what each operation does, and how the result looks like when applying it.

  GROUPBY 1 @category
  REDUCE COUNT 0 AS product_count
  REDUCE SUM 1 @price AS total_price
  REDUCE AVG 1 @rating AS avg_rating
  SORTBY 2 @total_price DESC
  LIMIT 0 10
```

## Multiple APPLY operations followed by GROUPBY and REDUCE

> `APPLY` can be used in various ways before and after `GROUPBY` and `REDUCE`.

Reviewer: ```suggestion
You can use `APPLY` in various ways before and after `GROUPBY` and `REDUCE`.
```

```sh
FT.AGGREGATE products "*"
  LOAD 3 @price @discount @quantity
  APPLY "@price - @discount" AS final_price
  APPLY "@final_price * @quantity" AS total_revenue
  GROUPBY 1 @category
  REDUCE SUM 1 @total_revenue AS total_category_revenue
  SORTBY 2 @total_category_revenue DESC
  LIMIT 0 10
```

## FILTER and PARAMS

> `FILTER` is used to remove unwanted records, while `PARAMS` allows parameterized queries.

Reviewer: There are no params in the query that follows this. Also, are `FILTER` and `PARAMS` closely related enough to be mentioned together? If this is based on the position of the keyword in the query then it might be better to say that explicitly.
```suggestion
Use `FILTER` to remove unwanted records, and `PARAMS` to pass values to parameterized queries.
```

```sh
FT.AGGREGATE products "*"
  LOAD 3 @price @rating @quantity
  FILTER "@price >= 500"
  FILTER "@rating >= 4.0"
  APPLY "@price * @quantity" AS total_value
  SORTBY 2 @total_value DESC
  LIMIT 0 10
  DIALECT 2
```

## Placement of FILTER before and after GROUPBY/APPLY

- **Before GROUPBY:** Removes records before aggregation.
- **After GROUPBY:** Filters based on aggregated results.
- **Before APPLY:** Ensures calculations are applied only to certain records.

> - **After APPLY:** Filters computed values.

Reviewer 2: Above you already have two such bullets, that have better description IMO and are enough for the before/after comparison.

## LOAD after GROUPBY/REDUCE

This is not allowed and you'll get a syntax error.

## Placement rules for specific parameters

| Parameter    | Placement                             |
|-----         |-----                                  |
| `TIMEOUT`    | Can be placed anywhere.               |
| `LIMIT`      | Must be at the end, before `DIALECT`. |
| `WITHCURSOR` | Must be at the end, before `DIALECT`. |
| `SCORER`     | Can be placed anywhere.               |
| `ADDSCORES`  | Must be before sorting.               |
| `DIALECT`    | Must be at the end.                   |

## LIMIT and WITHCURSOR used together

> In practical terms, `LIMIT` and `WITHCURSOR` are mutually exclusive. However, they can be used together.

Reviewer: It's not clear if you are saying that you can use them together advantageously or if one just overrides the other. (Or worse case, you get the mushroom cloud in the eastern sky.)

- When referencing fields loaded from documents. In the following example, `price` is a document field and must be prefixed with `@`.

```sh
FT.AGGREGATE products "*"
  LOAD 1 @price
  APPLY "@price * 1.1" AS adjusted_price
  SORTBY 2 @adjusted_price DESC
  LIMIT 0 10

 1) (integer) 200
 2) 1) "price"
    2) "623"
    3) "adjusted_price"
    4) "685.3"
 3) 1) "price"
    2) "619.75"
    3) "adjusted_price"
    4) "681.725"
  .
  .
  .
```

- When referencing fields inside a `FILTER` clause that were loaded from documents.

```sh
FT.AGGREGATE products "*"
  LOAD 1 @rating
  FILTER "@rating >= 4.5"
  LIMIT 0 10

 1) (integer) 5
 2) 1) "rating"
    2) "4.5"
 3) 1) "rating"
    2) "4.8"
 4) 1) "rating"
    2) "4.5"
  .
  .
  .
```

- When referencing fields inside `GROUPBY` or `REDUCE` clauses.

```sh
FT.AGGREGATE products "*"
  GROUPBY 1 @category
  REDUCE SUM 1 @price AS total_price
  LIMIT 0 10

1) (integer) 6
2) 1) "category"
   2) "Toys"
   3) "total_price"
   4) "9799.25"
3) 1) "category"
   2) "Electronics"
   3) "total_price"
   4) "10683.5"
4) 1) "category"
   2) "Apparel"
   3) "total_price"
   4) "10273.5"
  .
  .
  .
```

> - When referencing fields created by `REDUCE` in an `APPLY` or `FILTER` clauses.

Reviewer: Or alternatively, change "clauses" to "clause", whichever looks better.
```suggestion
- When referencing fields created by `REDUCE` in `APPLY` or `FILTER` clauses.
```

   4) "9799.25"
   5) "boosted_price"
   6) "11759.1"
3) 1) "category"
   2) "Electronics"
   3) "total_price"
   4) "10683.5"
   5) "boosted_price"
   6) "12820.2"
  .
  .
  .
```

- When referencing fields created by `APPLY` in another `APPLY` or `FILTER` clause.

```sh
FT.AGGREGATE products "*"
  LOAD 2 @price @discount
  APPLY "@price - @discount" AS net_price
  APPLY "@net_price * 1.1" AS marked_up
  FILTER "@net_price > 200"
  LIMIT 0 10

1) (integer) 60
2) 1) "price"
   2) "220"
   3) "discount"
   4) "0"
   5) "net_price"
   6) "220"
   7) "marked_up"
   8) "242"
3) 1) "price"
   2) "223.25"
   3) "discount"
   4) "1.5"
   5) "net_price"
   6) "221.75"
   7) "marked_up"
   8) "243.925"
  .
  .
  .
```

- When referencing fields created by `APPLY` in a `SORTBY` clause.

```sh
FT.AGGREGATE products "*"
  LOAD 2 @price @discount
  APPLY "@price - @discount" AS net_price
  SORTBY 2 @net_price DESC
  LIMIT 0 10

 1) (integer) 200
 2) 1) "price"
    2) "623"
    3) "discount"
    4) "6"
    5) "net_price"
    6) "617"
 3) 1) "price"
    2) "619.75"
    3) "discount"
    4) "4.5"
    5) "net_price"
    6) "615.25"
  .
  .
  .
```

## GROUPBY with multiple REDUCE operations

> `GROUPBY` can be followed by multiple `REDUCE` operations for different aggregations.

Reviewer: Not major, but the other similar example intros are active voice here.
```suggestion
You can use multiple `REDUCE` operations after `GROUPBY` for different aggregations.
```

    3) "rating"
    4) "4.2"
    5) "quantity"
    6) "10"
    7) "total_value"
    8) "6067.5"
 3) 1) "price"
    2) "541.75"
    3) "rating"
    4) "4.5"
    5) "quantity"
    6) "10"
    7) "total_value"
    8) "5417.5"
  .
  .
  .
```

## Placement of FILTER before and after GROUPBY/APPLY

- **Before GROUPBY:** Removes records before aggregation.
- **After GROUPBY:** Filters based on aggregated results.

## LOAD after GROUPBY/REDUCE

This is not allowed and you'll get a syntax error.

## Placement rules for specific parameters

| Parameter    | Placement                                                 |
|-----         |-----                                                      |
| `TIMEOUT`    | Can be placed anywhere.                                   |
| `LIMIT`      | Must be at the end, before `DIALECT`.                     |
| `WITHCURSOR` | Must be at the end, before `DIALECT`.                     |
| `SCORER`     | Must be placed between the query and pipeline operations. |
| `ADDSCORES`  | Must be placed between the query and pipeline operations. |
| `DIALECT`    | Must be at the end.                                       |

## LIMIT and WITHCURSOR used together

> While you wouldn't ordinarily use `LIMIT` and `WITHCURSOR` together in the same query, you can use them advantageously if doing so fits your workflow.

Reviewer: I'm not sure if this really captures what you meant, but I think it would be better to simplify this sentence a bit.
```suggestion
While you wouldn't ordinarily use `LIMIT` and `WITHCURSOR` together in the same query, you can do so if necessary.
```
